### DISCUSSION TITLE
RemainingDuration on Valve Service

### DISCUSSION THREAD

**Discussion by WindnSalt** [HomeSpan/HomeSpan#333] [Q&A]
For my garden I want to use this great library with an ESP32 C3 and some relays to replace the IDF implementation I have, as homespan is a lot more convenient. So far it works like a charme, but I do not understand one mechanism yet. From other implementations like homebridge or IDF I have seen, that homekit makes calls for getting the status of characteristics from time to time. For a valve that would mean in case the UI on my MacOS or iOS needs the remaining time, homebridge would call a callback for that characteristic to update the related value. In homespan I have seen that one can update a status through a loop in the service implementation. However, for remaining duration of a valve in an irrigation system, I would have to update the remaining duration every second. 
Is that the intended way to update this or am I missing something here?
Would that trigger a status update push from my device to homekit every second?

**Comment by HomeSpan** [comment]
The Remaining Duration Characteristic operates differently than most Characteristics since Apple automatically decrements the timer from within the Home App itself.  This saves the Accessory from needing to constantly update the timer every second, which is very inefficient.  However, there are two caveats:

First, the timer for each sprinkler head presented in the Home App is for display purposes only.  When the timer reaches zero the Home App doesn't actually do anything.  All of the control still resides in the Accessory, which must close the appropriate value AND inform the Home App that the valve has closed.  This means the Accessory still needs to keep track of the time regardless of what the Home App is doing, so it knows when to close the valve.  I believe the reason for this is that Apple recognizes that HomeKit connectivity is not 100% reliable - if something happened and the Home App could not signal the Accessory to turn off the water, the water could remain on indefinitely, which is definitely a problem.  By ensuring the Accessory is in charge of all timing, valves will close when they are supposed to even if your entire HomeKit system goes down.

Second, though the Home App is counting down on its own, this logic apparently operates ONLY while the Home App itself is open.  If you close the Home App and re-open it, the Home App will immediately request an update of all Characteristics from the Accessory, including the Remaining Duration for each sprinkler head, if that head is Active.   This is potentially a problem, because as noted above, the Accessory should not be sending updates to the Home App every second.

The solution is to update the Remaining Duration Characteristic every second from within the Accessory, but do so in a way that does NOT send that update to the Home App.  This is accomplished using the `setVal()` function with the optional second parameter set to `false` (see the API Reference for details).  In this fashion, the Remaining Duration Characteristic is always loaded with the correct time and is ready to provide it to the Home App upon request only when needed.

I wrote the sketch below a while back to test out this functionality but have not included it in any of the tutorials - I'll probably do so in a future update.   In any case, I just tested it again and it still operates as expected.   You'll note the sketch includes the ConfiguredName() Characteristic for each valve.  This is an undocumented feature in HAP-R2 that was added around the time Apple added the (undocumented) Television Services.  The purpose of this Characteristic is to allow the user to change the name of the Valves from within the Home App itself (as well enable/disable any given Valve).

```C++
/*********************************************************************************
 *  MIT License
 *  
 *  Copyright (c) 2021 Gregg E. Berman
 *  
 *  https://github.com/HomeSpan/HomeSpan
 *  
 *  Permission is hereby granted, free of charge, to any person obtaining a copy
 *  of this software and associated documentation files (the "Software"), to deal
 *  in the Software without restriction, including without limitation the rights
 *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *  copies of the Software, and to permit persons to whom the Software is
 *  furnished to do so, subject to the following conditions:
 *  
 *  The above copyright notice and this permission notice shall be included in all
 *  copies or substantial portions of the Software.
 *  
 *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 *  SOFTWARE.
 *  
 ********************************************************************************/

 
  ////////////////////////////////////////////////////////
  //                                                    //
  //    HomeSpan Reference SPRINKLERS Sketch:           //
  //                                                    //
  //    * IRRIGATION Service                            //
  //    * Multiple linked VALVE Services                //
  //                                                    //  
  ////////////////////////////////////////////////////////

 
#include "HomeSpan.h" 

struct Head : Service::Valve {

  SpanCharacteristic *active=new Characteristic::Active(0);
  SpanCharacteristic *inUse=new Characteristic::InUse(0);
  SpanCharacteristic *enabled = new Characteristic::IsConfigured(1);
  SpanCharacteristic *setDuration = new Characteristic::SetDuration(30);
  SpanCharacteristic *remainingDuration = new Characteristic::RemainingDuration(0);
  SpanCharacteristic *name;

  Head(const char *headName) : Service::Valve() {
    new Characteristic::ValveType(1);
    name=new Characteristic::ConfiguredName(headName);
    enabled->addPerms(PW);
  }

  boolean update() override {
    
    if(enabled->updated()){
      if(enabled->getNewVal()){
        Serial.printf("Head '%s' enabled\n",name->getString());
      } else {
        Serial.printf("Head '%s' disabled\n",name->getString());          
        if(active->getVal()){
          active->setVal(0);
          inUse->setVal(0);
          remainingDuration->setVal(0);
          Serial.printf("Head '%s' valve is closing\n",name->getString());          
        }
      }
    }

    if(active->updated()){
      if(active->getNewVal()){
        Serial.printf("Head '%s' valve is opening\n",name->getString());
        inUse->setVal(1);
        remainingDuration->setVal(setDuration->getVal());
      } else {
        Serial.printf("Head '%s' valve is closing\n",name->getString());
        inUse->setVal(0);
        remainingDuration->setVal(0);
      }
    }
    
    return(true);
  }

  void loop() override {
    if(active->getVal()){
      int remainingTime=setDuration->getVal()-active->timeVal()/1000;
         
      if(remainingTime<=0){
        Serial.printf("Head '%s' is closing (%d-second timer is complete)\n",name->getString(),setDuration->getVal());
        active->setVal(0);
        inUse->setVal(0);
        remainingDuration->setVal(0);
      } else

      if(remainingTime<remainingDuration->getVal()){
        remainingDuration->setVal(remainingTime,false);
      }
    }
  }

};

struct Sprinkler : Service::IrrigationSystem {

  SpanCharacteristic *active=new Characteristic::Active(0);
  SpanCharacteristic *programMode=new Characteristic::ProgramMode(0);
  
};

//////////////////////////////////////

void setup() {

  Serial.begin(115200);
  
  homeSpan.begin(Category::Sprinklers,"HomeSpan Sprinklers");

  new SpanAccessory();                                  
    new Service::AccessoryInformation();  
      new Characteristic::Identify();                           
                   
    (new Sprinkler())
      ->addLink(new Head("Head 1"))
      ->addLink(new Head("Head 2"))
      ->addLink(new Head("Head 3"))
      ->addLink(new Head("Head 4"))
      ;

} // end of setup()

//////////////////////////////////////

void loop(){
  
  homeSpan.poll();
  
} // end of loop()

//////////////////////////////////////
```

**Reply by WindnSalt** [reply]
Awesome! Exactly what I was looking for!

**Comment by carterscherer** [comment]
how are you supposed to allocate GPIO pins to the sprinkler heads?
